app/api.py

```diff
@@ -0,0 +1,195 @@
+import uvicorn
+from fastapi import FastAPI, File, UploadFile, HTTPException, Form, Request, Depends
+from fastapi.responses import JSONResponse
+from pydantic import BaseModel
+from typing import List, Dict, Any, Optional
+from contextlib import asynccontextmanager
+import shutil
+from fuzzywuzzy import fuzz, process
+from doctr.models import db_resnet50, crnn_vgg16_bn
+from app.doctr_module import load_custom_model, extract_names_and_boxes
+from app.pdf_ocr import extract_text_and_boxes_from_pdf
+from app.vector_database import store_text_in_vector_db
+from app.rag import retrieve_answer
+from config.config import logger
+from pathlib import Path
+import shutil
+
+
+app = FastAPI()
+
+# Define paths for the model weights
+det_weights_path = "models/db_resnet50-79bd7d70.pt"
+reco_weights_path = "models/crnn_vgg16_bn-9762b0b0.pt"
+
+# Global variables to hold the models
+det_model = None
+reco_model = None
+
+
+# Define a model for input data
+class NamePairs(BaseModel):
+    file_location: Optional[str] = None
+    names: Optional[list] = []
+
+
+# Custom dependency to handle form data and JSON
+async def get_name_pairs(
+    request: Request,
+    file_location: Optional[str] = Form(None),
+    names: Optional[str] = Form(None),
+    file: Optional[UploadFile] = File(None),
+) -> NamePairs:
+    content_type = request.headers.get("content-type")
+    if "application/json" in content_type:
+        try:
+            json_body = await request.json()
+            return NamePairs(**json_body)
+        except Exception as e:
+            raise HTTPException(status_code=400, detail="Invalid JSON input")
+    else:
+        # Handle form-data
+        # If a file is uploaded, use the file's location; otherwise, use the provided file location
+        if file:
+            temp_file_path = f"temp_{file.filename}"
+            with open(temp_file_path, "wb") as buffer:
```

Comment:
if you want to do it async you can implement aiofiles.open()

Reply:
Ordinary local file IO is blocking, and cannot easily and portably be made asynchronous. This means doing file IO may interfere with asyncio applications, which shouldn't block the executing thread. aiofiles helps with this by introducing asynchronous versions of files that support delegating operations to a separate thread pool.
```python
async with aiofiles.open('filename', mode='r') as f:
    contents = await f.read()
print(contents)
'My file contents'
```